File: basic/python/sort/selectionSort.py
# similar to bubble sort in that it works by repeatedly swapping items in a list.
# However, it's slightly more efficient than bubble sort because it only makes one swap per iteration.

def selection_sort(nums):
    n = len(nums)
    for i in range(n):
        min = i  # assume it is min and validate it
        for j in range(i+1, n):
            if nums[j] < nums[min]:
                min = j
        # Swap the found minimum element with the first element
        nums[i], nums[min] = nums[min], nums[i]
    return nums

File: basic/java/sort/SelectionSort.java
package java.sort;

public class SelectionSort {

   // time complexity: worst, average and best: O(n^2) ?
   // space complexity: O(1) (in-place sorting algorithm)
   // divide it into sorted and unsorted (implicitly)
   //
   // Sort by repeatedly finding the min in the unsorted array and swap it with the leftmost element
   // 
   public void sort(int[] arr) {
      int n = arr.length;
      for (int i = 0; i < n - 1; i++) {
         int min = i; // assume it is min and validate it by comparing with the rest, correct the min if found
         
         for (int j = i + 1; j < n; j++) { // compare, starting from the next one
            if (arr[j] < arr[min]) {
               min = j; // update the min value
            }
         }
         // swapping arr[min] and arr[i]
         int temp = arr[min];
         arr[min] = arr[i];
         arr[i] = temp;
      }
   }

   public void printArray(int[] arr) {
      int n = arr.length;
      for (int i = 0; i < n; i++) {
         System.out.print(arr[i] + " ");
      }
      System.out.println();
   }
   public static void main(String[] args) {
      int[] arr = new int[] { 5, 1, 2, 9, 10 };
      SelectionSort ss = new SelectionSort();
      ss.printArray(arr);
      ss.sort(arr);
      ss.printArray(arr);
   }

}